File: Sources/XMTPiOS/Frames/FramesTypes.swift
//
//  File.swift
//  
//
//  Created by Alex Risch on 3/28/24.
//

import Foundation

typealias AcceptedFrameClients = [String: String]

enum OpenFrameButton: Codable {
    case link(target: String, label: String)
    case mint(target: String, label: String)
    case post(target: String?, label: String)
    case postRedirect(target: String?, label: String)

    enum CodingKeys: CodingKey {
        case action, target, label
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let action = try container.decode(String.self, forKey: .action)
        guard let target = try container.decodeIfPresent(String.self, forKey: .target) else {
            throw FramesClientError.missingTarget
        }
        let label = try container.decode(String.self, forKey: .label)

        switch action {
        case "link":
            self = .link(target: target, label: label)
        case "mint":
            self = .mint(target: target, label: label)
        case "post":
            self = .post(target: target, label: label)
        case "post_redirect":
            self = .postRedirect(target: target, label: label)
        default:
            throw DecodingError.dataCorruptedError(forKey: .action, in: container, debugDescription: "Invalid action value")
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .link(let target, let label):
            try container.encode("link", forKey: .action)
            try container.encode(target, forKey: .target)
            try container.encode(label, forKey: .label)
        case .mint(let target, let label):
            try container.encode("mint", forKey: .action)
            try container.encode(target, forKey: .target)
            try container.encode(label, forKey: .label)
        case .post(let target, let label):
            try container.encode("post", forKey: .action)
            try container.encode(target, forKey: .target)
            try container.encode(label, forKey: .label)
        case .postRedirect(let target, let label):
            try container.encode("post_redirect", forKey: .action)
            try container.encode(target, forKey: .target)
            try container.encode(label, forKey: .label)
        }
    }
}

public struct OpenFrameImage: Codable {
    let content: String
    let aspectRatio: AspectRatio?
    let alt: String?
}

public enum AspectRatio: String, Codable {
    case ratio_1_91_1 = "1.91.1"
    case ratio_1_1 = "1:1"
}

public struct TextInput: Codable {
    let content: String
}

struct OpenFrameResult: Codable {
    let acceptedClients: AcceptedFrameClients
    let image: OpenFrameImage
    let postUrl: String?
    let textInput: TextInput?
    let buttons: [String: OpenFrameButton]?
    let ogImage: String
    let state: String?
};

public struct GetMetadataResponse: Codable {
    let url: String
    public let extractedTags: [String: String]
}

public struct PostRedirectResponse: Codable  {
    let originalUrl: String
    let redirectedTo: String
};

public struct OpenFramesUntrustedData: Codable {
    let url: String
        let timestamp: Int
        let buttonIndex: Int
        let inputText: String?
        let state: String?
}

public typealias FramesApiRedirectResponse = PostRedirectResponse;

public struct FramePostUntrustedData: Codable {
    let url: String
    let timestamp: UInt64
    let buttonIndex: Int32
    let inputText: String?
    let state: String?
    let walletAddress: String
    let opaqueConversationIdentifier: String
    let unixTimestamp: UInt32
}

public struct FramePostTrustedData: Codable {
    let messageBytes: String
}

public struct FramePostPayload: Codable {
    let clientProtocol: String
    let untrustedData: FramePostUntrustedData
    let trustedData: FramePostTrustedData
}

public struct DmActionInputs: Codable {
    public let conversationTopic: String?
    public let participantAccountAddresses: [String]
    public init(conversationTopic: String? = nil, participantAccountAddresses: [String]) {
        self.conversationTopic = conversationTopic
        self.participantAccountAddresses = participantAccountAddresses
    }
}

public struct GroupActionInputs: Codable {
    let groupId: Data
    let groupSecret: Data
}

public enum ConversationActionInputs: Codable {
    case dm(DmActionInputs)
    case group(GroupActionInputs)
}

public struct FrameActionInputs: Codable {
    let frameUrl: String
    let buttonIndex: Int32
    let inputText: String?
    let state: String?
    let conversationInputs: ConversationActionInputs
    public init(frameUrl: String, buttonIndex: Int32, inputText: String?, state: String?, conversationInputs: ConversationActionInputs) {
        self.frameUrl = frameUrl
        self.buttonIndex = buttonIndex
        self.inputText = inputText
        self.state = state
        self.conversationInputs = conversationInputs
    }
}

File: library/src/main/java/org/xmtp/android/library/frames/FramesTypes.kt
package org.xmtp.android.library.frames

typealias AcceptedFrameClients = Map<String, String>

sealed class OpenFrameButton {
    abstract val target: String?
    abstract val label: String

    data class Link(override val target: String, override val label: String) : OpenFrameButton()

    data class Mint(override val target: String, override val label: String) : OpenFrameButton()

    data class Post(override val target: String?, override val label: String) : OpenFrameButton()

    data class PostRedirect(override val target: String?, override val label: String) : OpenFrameButton()
}

data class OpenFrameImage(
    val content: String,
    val aspectRatio: AspectRatio?,
    val alt: String?
)

enum class AspectRatio(val ratio: String) {
    RATIO_1_91_1("1.91.1"),
    RATIO_1_1("1:1")
}

data class TextInput(val content: String)

data class OpenFrameResult(
    val acceptedClients: AcceptedFrameClients,
    val image: OpenFrameImage,
    val postUrl: String?,
    val textInput: TextInput?,
    val buttons: Map<String, OpenFrameButton>?,
    val ogImage: String,
    val state: String?
)

data class GetMetadataResponse(
    val url: String,
    val extractedTags: Map<String, String>
)

data class PostRedirectResponse(
    val originalUrl: String,
    val redirectedTo: String
)

data class OpenFramesUntrustedData(
    val url: String,
    val timestamp: Int,
    val buttonIndex: Int,
    val inputText: String?,
    val state: String?
)

typealias FramesApiRedirectResponse = PostRedirectResponse

data class FramePostUntrustedData(
    val url: String,
    val timestamp: Long,
    val buttonIndex: Int,
    val inputText: String?,
    val state: String?,
    val walletAddress: String,
    val opaqueConversationIdentifier: String,
    val unixTimestamp: Int
)

data class FramePostTrustedData(
    val messageBytes: String
)

data class FramePostPayload(
    val clientProtocol: String,
    val untrustedData: FramePostUntrustedData,
    val trustedData: FramePostTrustedData
)

data class DmActionInputs(
    val conversationTopic: String?,
    val participantAccountAddresses: List<String>
)

data class GroupActionInputs(
    val groupId: ByteArray,
    val groupSecret: ByteArray
)

sealed class ConversationActionInputs {
    data class Dm(val inputs: DmActionInputs) : ConversationActionInputs()
    data class Group(val inputs: GroupActionInputs) : ConversationActionInputs()
}

data class FrameActionInputs(
    val frameUrl: String,
    val buttonIndex: Int,
    val inputText: String?,
    val state: String?,
    val conversationInputs: ConversationActionInputs
)
